checkpoint/orbax/checkpoint/_src/serialization/replica_slices.py

```diff
@@ -0,0 +1,198 @@
+# Copyright 2024 The Orbax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+import dataclasses
+
+from absl import logging
+import jax
+import jax.numpy as jnp
+import math
+import numpy as np
+from orbax.checkpoint._src.arrays import numpy_utils
+from orbax.checkpoint._src.arrays import types
+from orbax.checkpoint._src.arrays.fragments import Fragment, Fragments, validate_fragments_can_be_stacked
+from orbax.checkpoint._src.multihost import multihost
+from typing import Callable, Optional, Sequence
+
+
+Shape = types.Shape
+Index = types.Index
+
+
+@dataclasses.dataclass(frozen=True)
+class ReplicaSliceOnDevice:
+  """
+  ReplicaSliceOnDevice represents the part of a jax.Shard that a replica is
+  uniquely responsible for.
+
+  With single-replica checkpointing the entirety of each jax.Shard is owned by
+  exactly one replica. (Currently the only option.)
+  """
+
+  replica_id: int
+  index: Index
+  data: jax.Array
+
+
+@dataclasses.dataclass
+class ReplicaSlices:
+  """
+  ReplicaSlices groups all the sliced data of one jax.Array that a replica is
+  uniquely responsible for. Slices may be either on-device (as a list of
+  ReplicaSliceOnDevice) or on-host (as a list of indices and numpy arrays).
+  """
+
+  global_shape: Shape
+  local_shape: Shape
+  sharding: jax.sharding.Sharding
+  dtype: np.dtype
+  # Whether the replica slices have been transferred and are ready as ndarrays
+  transferred: bool
```

Comment:
Is this really necessary if we can just check `replica_slices` to see whether it contains numpy arrays or not?

Reply:
Because there might be no replica_slices (e.g. because we are in single-replica mode and the current process has no shards with `shard.replica_id == replica_id`)?